lang/syn/src/codegen/accounts/try_accounts.rs

```diff
@@ -32,14 +32,33 @@ pub fn generate(accs: &AccountsStruct) -> proc_macro2::TokenStream {
                     // `init` and `zero` acccounts are special cased as they are
                     // deserialized by constraints. Here, we just take out the
                     // AccountInfo for later use at constraint validation time.
-                    if is_init(af) || f.constraints.zeroed.is_some() {
+                    if is_init(af) || f.constraints.zeroed.is_some()  {
                         let name = &f.ident;
-                        quote!{
-                            if accounts.is_empty() {
-                                return Err(anchor_lang::error::ErrorCode::AccountNotEnoughKeys.into());
+                        // Optional accounts have slightly different behavior here and
+                        // we can't leverage the try_accounts implementation for zero and init.
+                        if f.is_optional {
+                            // Thus, this block essentially reimplements the try_accounts 
+                            // behavior with optional accounts minus the deserialziation.
+                            quote! {
+                                let #name = if accounts.is_empty() {
```

Comment:
Reminder to myself to come back to this. Why does this double up the logic on the `try_accounts` inside optional_accounts. 

Probably because there are constraints to be run on the account then you have to load it up first as an account info before doing try_accounts? Or maybe with `Account<'info, T>` types you just never actually call `try_accounts`?

Still need to get rid of `if account.is_empty() return None` behaviour

Reply:
It seems like this behaviour is linked to whether an account is expected be initialized or not. So `zero` and `init` will both lazily load the account info and perform the checks / initialization needed on the account before loading it as the correct type.